File: forgerock-auth/src/main/java/org/forgerock/android/auth/callback/PasswordCallback.java
/*
 * Copyright (c) 2019 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

package org.forgerock.android.auth.callback;

import androidx.annotation.Keep;

import org.json.JSONObject;

import java.util.Arrays;

import lombok.NoArgsConstructor;

/**
 * Callback to collect a username
 */
public class PasswordCallback extends AbstractPromptCallback {

    @Keep
    public PasswordCallback() {
    }

    @Keep
    public PasswordCallback(JSONObject jsonObject, int index) {
        super(jsonObject, index);
    }

    /**
     * Set the retrieved password.
     *
     * <p>
     *
     * @param password the retrieved password, which may be null.
     *
     */
    public void setPassword(char[] password) {
        setValue(new String(password));
        Arrays.fill(password, ' ');
    }

    @Override
    public String getType() {
        return "PasswordCallback";
    }
}

File: e2e/app/src/main/java/com/example/app/callback/PasswordCallback.kt
/*
 * Copyright (c) 2023 ForgeRock. All rights reserved.
 *
 *  This software may be modified and distributed under the terms
 *  of the MIT license. See the LICENSE file for details.
 */

package com.example.app.callback

import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Visibility
import androidx.compose.material.icons.filled.VisibilityOff
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.text.input.PasswordVisualTransformation
import androidx.compose.ui.text.input.VisualTransformation
import androidx.compose.ui.unit.dp
import org.forgerock.android.auth.callback.PasswordCallback

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun PasswordCallback(callback: PasswordCallback) {

    Row(modifier = Modifier.padding(4.dp)) {
        var password by rememberSaveable { mutableStateOf("") }
        var passwordVisibility by remember { mutableStateOf(false) }

        OutlinedTextField(
            value = password,
            onValueChange = { value ->
                password = value
                callback.setPassword(password.toCharArray())
            },
            label = { Text(callback.prompt) },
            trailingIcon = {
                IconButton(onClick = { passwordVisibility = !passwordVisibility }) {
                    if (passwordVisibility)
                        Icon(Icons.Filled.Visibility, contentDescription = null)
                    else
                        Icon(Icons.Filled.VisibilityOff, contentDescription = null)
                }
            },
            keyboardOptions = KeyboardOptions(
                keyboardType = KeyboardType.Password
            ),
            visualTransformation = if (passwordVisibility) VisualTransformation.None
            else PasswordVisualTransformation()
        )
    }

}